### PR TITLE
[move-2024][easy] Disallow macro variables to be used as exp dotted base expressions

### PR DESCRIPTION
**Pull request by cgswords**
## Description 

What it says on the tin -- this is just to reduce programmer confusion

## Test Plan 

New test case plus updated expectations

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_duck_typing_method.exp [hunk 1]
```diff
@@ -0,0 +1,9 @@
+error[E03006]: unexpected name in this position
+   ┌─ tests/move_2024/typing/macro_duck_typing_method.move:18:9
+   │
+18 │         $x.foo()
+   │         ^^ Macro parameters are not allowed to appear in paths
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_duck_typing_method.exp:5]: Oh wait, this ain't good. Or is it? Thoughts here?

Maybe we should suggest they first let-bind the argument, e.g. "Try binding the argument to a local variable first, e.g. 'let x = $x;'

Comment (tnowacki) [reply external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_duck_typing_method.exp:5]: @cgswords thought on adding this to the note?

File: external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_unique_use_fun_scopes_in_lambda_arg.exp [hunk 2]
```diff
@@ -0,0 +1,10 @@
+error[E03006]: unexpected name in this position
+  ┌─ tests/move_2024/typing/macro_unique_use_fun_scopes_in_lambda_arg.move:7:12
+  │
+7 │         $f($x.id())
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_unique_use_fun_scopes_in_lambda_arg.exp:4]: Could we fix this test

File: external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_unique_use_fun_scopes_captured_lambda.exp [hunk 3]
```diff
@@ -0,0 +1,10 @@
+error[E03006]: unexpected name in this position
+  ┌─ tests/move_2024/typing/macro_unique_use_fun_scopes_captured_lambda.move:7:12
+  │
+7 │         $f($x.into())
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_unique_use_fun_scopes_captured_lambda.exp:4]: ditto

File: external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_duck_typing_method_invalid.exp [hunk 4]
```diff
@@ -1,3 +1,13 @@
+error[E03006]: unexpected name in this position
+  ┌─ tests/move_2024/typing/macro_duck_typing_method_invalid.move:4:9
+  │
+4 │         $x.foo()
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_duck_typing_method_invalid.exp:4]: ditto

File: external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_arg_by_name_invalid_usage_var.exp [hunk 5]
```diff
@@ -4,9 +4,29 @@ error[E04026]: invalid 'copy' usage
...
+  │              ^^ Macro parameters are not allowed to appear in paths
+  │
+  = To use a macro parameter as a value in a path expression, first bind it to a local variable, e.g. 'let x = $x;'
+  = Macro parameters are always treated as value expressions, and are not modified by path operations.
+    Path operations include 'move', 'copy', '&', '&mut', and field references
+
 error[E04027]: invalid 'move' usage
   ┌─ tests/move_2024/typing/macro_arg_by_name_invalid_usage_var.move:5:9
   │
 5 │         move $x;
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_arg_by_name_invalid_usage_var.exp:20]: Any way to not duplicate these errors like this?

File: external-crates/move/crates/move-compiler/src/naming/translate.rs [hunk 6]
```diff
@@ -1817,6 +1817,24 @@ fn dotted(context: &mut Context, edot: E::ExpDotted) -> Option<N::ExpDotted> {
...
+                            it to a local variable, e.g. 'let {0} = ${0};'",
+                        &n.value.name.to_string()[1..]
+                    ));
+                    diag.add_note(
+                        "Macro parameters are always treated as value expressions, and are not \
+                        modified by path operations.\n\
+                        Path operations include 'move', 'copy', '&', '&mut', and field references",
+                    );
+                    context.env.add_diag(diag);
+                    N::ExpDotted_::Exp(ne)
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/naming/translate.rs:1836]: Maybe this should be UnresolvedError?

**Review by tnowacki — CHANGES_REQUESTED** [review]

File: external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_unique_use_fun_scopes_in_lambda_arg.exp (hunk 2 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_unique_use_fun_scopes_in_lambda_arg.exp]: Fixed all of these

File: external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_arg_by_name_invalid_usage_value.exp [hunk 7]
```diff
@@ -30,3 +18,11 @@ error[E03006]: unexpected name in this position
   = Macro parameters are always treated as value expressions, and are not modified by path operations.
     Path operations include 'move', 'copy', '&', '&mut', and field references
 
+warning[W09005]: dead or unreachable code
+  ┌─ tests/move_2024/typing/macro_arg_by_name_invalid_usage_value.move:9:14
+  │
+9 │         foo!(0);
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_arg_by_name_invalid_usage_value.exp:24]: intentional?

File: external-crates/move/crates/move-compiler/src/naming/translate.rs [hunk 8]
```diff
@@ -1833,7 +1833,7 @@ fn dotted(context: &mut Context, edot: E::ExpDotted) -> Option<N::ExpDotted> {
                         Path operations include 'move', 'copy', '&', '&mut', and field references",
                     );
                     context.env.add_diag(diag);
-                    N::ExpDotted_::Exp(ne)
+                    N::ExpDotted_::Exp(Box::new(sp(ne.loc, N::Exp_::UnresolvedError)))
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/naming/translate.rs:1836]: 🎉

**Review by tnowacki — APPROVED** [review]

File: external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_arg_by_name_invalid_usage_value.exp (hunk 7 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/tests/move_2024/typing/macro_arg_by_name_invalid_usage_value.exp]: No, the error thing caused it to shift. Thanks for catching it!

**Pull request merged**